File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.java
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        PriorityQueue<Long> result = new PriorityQueue<>();
        Queue<TreeNode> que = new ArrayDeque<>();
        que.offer(root);

        while (!que.isEmpty()) {
            int n = que.size();
            long levelSum = 0L;

            for (int i = 0; i < n; i++) {
                TreeNode node = que.peek();
                levelSum += node.val;
                que.poll();

                if (node.left != null) {
                    que.offer(node.left);
                }
                if (node.right != null) {
                    que.offer(node.right);
                }
            }

            result.offer(levelSum);

            if (result.size() > k) {
                result.poll();
            }
        }

        if (result.size() < k) return -1;

        return result.peek();
    }
}

File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.cpp
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        queue<TreeNode*> que;
        que.push(root);

        priority_queue<long long, vector<long long>, greater<long long>> result;

        while (!que.empty()) {
            int n = que.size();
            long long levelSum = 0;

            for (int i = 0; i < n; i++) {
                TreeNode* node = que.front();
                levelSum += node->val;
                que.pop();

                if (node->left != NULL) {
                    que.push(node->left);
                }
                if (node->right != NULL) {
                    que.push(node->right);
                }
            }

            result.push(levelSum);
            if (result.size() > k) {
                result.pop();
            }
        }

        if (result.size() < k) return -1;

        return result.top();
    }
};